File: Solutions/948_Bag_of_Tokens/948_Bag_of_Tokens.py
class Solution:
    def bagOfTokensScore(self, tokens, P):
        tokens.sort()
        beg, end, S, max_S = 0, len(tokens) - 1, 0, 0
        while beg <= end:
            if tokens[beg] <= P:
                P -= tokens[beg]
                S += 1
                max_S = max(max_S, S)
                beg += 1
            elif S >= 1:
                P += tokens[end]
                S -= 1
                end -= 1
            else: break
        return max_S

#Third fastest
class Solution:
    def bagOfTokensScore(self, tokens: List[int], P: int) -> int:
        tokens,l,r,points,ans=sorted(tokens),0,len(tokens)-1,0,0
        #print(tokens)
        while l<=r:
            if tokens[l]<=P:
                P-=tokens[l]
                points+=1
                l+=1
            elif points>=1:
                points-=1
                P+=tokens[r]
                r-=1
            else:
                break
            ans=max(ans,points)
        return ans

#Second fastest
from collections import deque

class Solution:
    def bagOfTokensScore(self, tokens: List[int], P: int) -> int:
        tokens = sorted(tokens)
        tokens = deque(tokens)
        score = 0
        power = P
        while tokens:
            if power >= tokens[0]:
                power -= tokens[0]
                score += 1
                tokens.popleft()
            elif score > 0 and len(tokens) > 1 and tokens[-1] >= tokens[0]:
                power += tokens[-1]
                score -= 1
                tokens.pop()
            else:
                tokens.pop()
        return score

#Fastest
class Solution:
    def bagOfTokensScore(self, tokens: List[int], P: int) -> int:
        tokens.sort()
        if not tokens or P < tokens[0]:
            return 0
        score = 0
        left, right = 0, len(tokens) - 1
        while left <= right:
            if P >= tokens[left]:
                P -= tokens[left]
                left += 1
                score += 1
            else:
                if right - left > 1:
                    P += tokens[right]
                    right -= 1
                    score -= 1
                else:
                    break
        return score

File: Solutions/948_Bag_of_Tokens/948_Bag_of_Tokens.cs
public class Solution {
    public int BagOfTokensScore(int[] tokens, int P) {
        Array.Sort(tokens);
        int beg = 0, end = tokens.Length - 1, s = 0, max_s = 0;
        while(beg <= end) {
            if(tokens[beg] <= P) {
                P -= tokens[beg];
                s += 1;
                max_s = Math.Max(max_s, s);
                beg += 1;
            }
            else if(s >= 1){
                P += tokens[end];
                s -= 1;
                end -= 1;
            }
            else
                break;
        }
        
        return max_s;
    }
}

//Third fastest
public class Solution {
    public int BagOfTokensScore(int[] tokens, int P) {
     
        Array.Sort(tokens);
        var l=0;
        var r=tokens.Length-1;
        var points=0;
        var max=0;
        while(l<=r){
            if(P>=tokens[l]){
                points++;
                P-=tokens[l++];
            }else if(points>0){
                max = Math.Max(max, points--);
                P += tokens[r--];
            }else{
                break;
            }
        }
        max = Math.Max(max, points);
        return max;
    }
    
}

//Second fastest
public class Solution {
    // Approach: Greedy. Sort the array and consider low token power for tokenUp and high for losing the points to gain max ponits.
    public int BagOfTokensScore(int[] tokens, int P) {
        Array.Sort(tokens);
        int l = 0, r = tokens.Length - 1, ans = 0, points = 0;
        while(l <= r && (P >= tokens[l] || points > 0))
        {
            while(l <= r && P >= tokens[l])
            {
                P -= tokens[l++];
                points++;
            }
            ans = Math.Max(ans, points);
            if (l <= r && points > 0)
            {
                P += tokens[r--];
                points--;
            }
        }
        
        return ans;
    }
}

//Fastest
public class Solution {
    public int BagOfTokensScore(int[] tokens, int P) {
        Array.Sort(tokens);
			int lo = 0, hi = tokens.Length - 1;
			int points = 0, ans = 0;
			while (lo <= hi && (P >= tokens[lo] || points > 0))
			{
				while (lo <= hi && P >= tokens[lo])
				{
					P -= tokens[lo++];
					points++;
				}
				ans = Math.Max(ans, points);
				if (lo <= hi && points > 0)
				{
					P += tokens[hi--];
					points--;
				}
			}
			return ans;
    }
}